builtin/rebase.c

```diff
@@ -0,0 +1,687 @@
+
+	strbuf_reset(&buf);
+	if (read_one(state_dir_path("quiet", opts), &buf))
+		return -1;
+	if (buf.len)
+		opts->flags &= ~REBASE_VERBOSE;
+	else
+		opts->flags |= REBASE_VERBOSE;
+
+	if (file_exists(state_dir_path("verbose", opts)))
+		opts->flags |= REBASE_VERBOSE;
+
+	strbuf_reset(&strategy);
+	if (file_exists(state_dir_path ("strategy", opts))) {
+		if (read_one(state_dir_path ("strategy", opts), &strategy))
+			return -1;
+		opts->strategy = strategy.buf;
+	}
+
+	strbuf_release(&buf);
+
+	return 0;
+}
+
+static int finish_rebase(struct rebase_options *opts)
+{
+	struct strbuf dir = STRBUF_INIT;
+	const char *argv_gc_auto[] = { "gc", "--auto", NULL };
+
+	delete_ref(NULL, "REBASE_HEAD", NULL, REF_NO_DEREF);
+	apply_autostash();
+	close_all_packs(the_repository->objects);
+	/*
+	 * We ignore errors in 'gc --auto', since the
+	 * user should see them.
+	 */
+	run_command_v_opt(argv_gc_auto, RUN_GIT_CMD);
+	strbuf_addstr(&dir, opts->state_dir);
+	remove_dir_recursively(&dir, 0);
+	strbuf_release(&dir);
+
+	return 0;
+}
+
+static struct commit *peel_committish(const char *name)
+{
+	struct object *obj;
+	struct object_id oid;
+
+	if (get_oid(name, &oid))
+		return NULL;
+	obj = parse_object(&oid);
+	return (struct commit *)peel_to_type(name, 0, obj, OBJ_COMMIT);
+}
+
+static void add_var(struct strbuf *buf, const char *name, const char *value)
+{
+	strbuf_addstr(buf, name);
+	strbuf_addstr(buf, "=");
+	sq_quote_buf(buf, value);
+	strbuf_addstr(buf, "; ");
+}
+
+static int run_specific_rebase(struct rebase_options *opts)
+{
+	const char *argv[] = { NULL, NULL };
+	struct strbuf script_snippet = STRBUF_INIT;
+	int status;
+	const char *backend, *backend_func;
+
+	add_var(&script_snippet, "GIT_DIR", absolute_path(get_git_dir()));
+	add_var(&script_snippet, "state_dir", opts->state_dir);
+
+	if (opts->upstream_name)
+		add_var(&script_snippet, "upstream_name", opts->upstream_name);
+	if (opts->upstream)
+		add_var(&script_snippet, "upstream",
+			oid_to_hex(&opts->upstream->object.oid));
+	add_var(&script_snippet, "head_name", opts->head_name);
+	add_var(&script_snippet, "orig_head", oid_to_hex(&opts->orig_head));
+	add_var(&script_snippet, "onto", oid_to_hex(&opts->onto->object.oid));
+	if (opts->onto_name)
+		add_var(&script_snippet, "onto_name", opts->onto_name);
+	if (opts->revisions)
+		add_var(&script_snippet, "revisions", opts->revisions);
+	add_var(&script_snippet, "GIT_QUIET",
+		opts->flags & REBASE_VERBOSE ? "" : "t");
+	if (opts->git_am_opt)
+		add_var(&script_snippet, "git_am_opt", opts->git_am_opt);
+	add_var(&script_snippet, "verbose",
+		opts->flags & REBASE_VERBOSE ? "t" : "");
+	add_var(&script_snippet, "diffstat",
+		opts->flags & REBASE_DIFFSTAT ? "t" : "");
+	add_var(&script_snippet, "action", opts->action ? opts->action : "");
+	if (opts->strategy)
+		add_var(&script_snippet, "strategy", opts->strategy);
+	add_var(&script_snippet, "do_merge",
+		opts->flags & REBASE_DO_MERGE ? "t" : "");
```

Comment:
Please do keep https://github.com/git/git/pull/505#discussion_r200508259 in mind. `do_merge` is not used by *any* of the rebase backends, therefore this `add_var` is unnecessary.

Further, I would like to encourage you to keep working on only the subset of options (or for that matter, environment variables listed in https://github.com/git/git/pull/505#discussion_r200508259) that are relevant to the `am` backend, until that one *works*.

(And the `do_merge` variable is squarely in the `--merge` backend square.)